### DISCUSSION TITLE
Setting up poetry `.venv`

### DISCUSSION THREAD

**Discussion by Anselmoo** [community/community#17462] [Codespaces]
Dear Community,

I am aware of the [code.visualstudio.com - devcontainerjson-reference)](https://code.visualstudio.com/docs/remote/devcontainerjson-reference), however, I was not able to get it running:

I thought that `postCreateCommand` supports only `str` and **not** `list`, so I choose the following command for the `devcontainer.json`

```json
// Use 'postCreateCommand' to run commands after the container is created.
	"postCreateCommand": "pip install poetry && poetry install & poetry shell",
```

The image is built with the errors:

```log
✔ Finishing up...
✖ Running postCreateCommand...
✔ Done
```
And the error is the following:

```log
Error: Command failed: /bin/sh -c pip install poetry && poetry install -E all && poetry shell
at mc (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:100:137)
at processTicksAndRejections (internal/process/task_queues.js:95:5)
at async pc (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:93:3564)
at async dc (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:93:2839)
at async JR (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:226:21572)
at async KR (/usr/lib/node_modules/@microsoft/vscode-dev-containers-cli/dist/spec-node/devContainersSpecCLI.js:226:19769)
```

Did anybody know, how to setup a proper `poetry` environment in __Codespaces__, please?

thx

**Comment by ronaldgrn** [comment]
In case anyone else comes across this, you can setup poetry by adding it as a feature to your devcontainer.json. eg.

```json
{
	"name": "Python 3",
	"image": "mcr.microsoft.com/devcontainers/python:1-3.9-bookworm",
	"features": {
		"ghcr.io/devcontainers-contrib/features/poetry:2": {
			"version": "latest"
		}
	},
```

If you want to run commands using poetry during container setup then use poetry run. eg.
```json
	"postCreateCommand": "poetry run python manage.py migrate",
```

Instead of `poetry shell` I'd recommend setting up your interpreter path when you know it
```json
	"customizations": {
		"vscode": {
			"settings": {
				"python.defaultInterpreterPath": "~/.cache/pypoetry/virtualenvs/xyz--py3.9/bin/python"
			}
		}
	}
```

**Reply by juhoinkinen** [reply]
Thanks @ronaldgrn, this was the only way I got the environment to also _activate_ automatically on startup.

First I was unwilling to hardcode the path, which includes a hash, which seemed random. But in a [realpython.com article](https://realpython.com/dependency-management-python-poetry/#use-poetrys-virtual-environments) it is explained:

> As you can see, Poetry constructed a unique name for your project’s virtual environment. It contains your package name from pyproject.toml and the corresponding Python version. *The seemingly random string in the middle, Dod5cRxq, is a [Base64-encoded](https://en.wikipedia.org/wiki/Base64) hash value of the path leading up to your project’s parent directory. It ties the name of a virtual environment to your project’s location on disk.*

But actually better way could be make Poetry [create the virtualenv inside the project’s root directory](https://python-poetry.org/docs/configuration/#virtualenvsin-project).
